packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionHierarchy_def.hpp

```diff
@@ -280,41 +250,73 @@ void MakeCoarseLevelMaps2(const int maxRegPerGID,
-                                      8*regMatrices[maxLevel][0]->getCrsGraph()->getNodeMaxNumRowEntries(),
+                                      8*regMatrix[0]->getCrsGraph()->getNodeMaxNumRowEntries(),
                                       Xpetra::DynamicProfile);
-  //      coarseCompOp->setAllToScalar(SC_ZERO);
-  //      coarseCompOp->describe(*fos, Teuchos::VERB_EXTREME);
 
-  regionalToComposite(regMatrices[maxLevel],
-                      quasiRegRowMaps[maxLevel], quasiRegColMaps[maxLevel],
-                      regRowImporters[maxLevel], Xpetra::ADD,
+  regionalToComposite(regMatrix,
+                      quasiRegRowMap, quasiRegColMap,
+                      regRowImporter, Xpetra::ADD,
                       coarseCompOp);
 
-  //      coarseCompOp->fillComplete(compRowMaps[maxLevel], compRowMaps[maxLevel]);
-  //      TEUCHOS_ASSERT(coarseCompOp->isFillComplete());
-  //
-  //      sleep(1);
-  //      std::cout << myRank << " | Printing coarseCompOp ..." << std::endl;
-  //      Comm->barrier();
-  //      coarseCompOp->describe(*fos, Teuchos::VERB_HIGH);
+  const int dofsPerNode = regMatrix[0]->getRowMap()->getNodeNumElements() / regCoarseCoordinates[0]->getMap()->getNodeNumElements();
+  for(int regIdx = 0; regIdx < maxRegPerProc; ++regIdx) {
+    const int check       = regMatrix[regIdx]->getRowMap()->getNodeNumElements()
+      % regCoarseCoordinates[regIdx]->getMap()->getNodeNumElements();
+    TEUCHOS_ASSERT(check == 0);
+  }
+
+  RCP<Map> compCoordMap;
+  std::vector<RCP<Xpetra::Import<LocalOrdinal, GlobalOrdinal, Node> > > regCoordImporter(maxRegPerProc);
+  if(dofsPerNode == 1) {
+    compCoordMap = compRowMap;
+    for(int regionIdx = 0; regionIdx < maxRegPerProc; ++regionIdx) {
+      regCoordImporter[regionIdx] = regRowImporter[regionIdx];
+    }
+  } else {
+    using size_type = typename Teuchos::Array<GlobalOrdinal>::size_type;
+    Array<GlobalOrdinal> compCoordMapData(compRowMap->getNodeNumElements() / dofsPerNode);
+    ArrayView<const GlobalOrdinal> compRowMapData = compRowMap->getNodeElementList();
+    for(size_type nodeIdx = 0; nodeIdx < compCoordMapData.size(); ++nodeIdx) {
+      compCoordMapData[nodeIdx] = compRowMapData[nodeIdx*dofsPerNode] / dofsPerNode;
+    }
+    compCoordMap = MapFactory::Build(compRowMap->lib(),
+                                     compRowMap->getGlobalNumElements() / dofsPerNode,
+                                     compCoordMapData(),
+                                     compRowMap->getIndexBase(),
+                                     compRowMap->getComm());
+  }
+  compCoarseCoordinates = MultiVectorFactory::Build(compCoordMap, regCoarseCoordinates[0]->getNumVectors());
+  TEUCHOS_ASSERT(Teuchos::nonnull(compCoarseCoordinates));
+
+  // The following looks like regionalToComposite for Vector
+  // but it is a bit different since we do not want to add
+  // entries in the coordinate MultiVector as we would for
+  // a solution or residual vector.
```

Comment:
We could add an argument `Xpetra::CombineMode combineMode` defaulting to `Xpetra::ADD` to `regionalToComposite()`. Then, we can call it here with a combine mode `Xpetra::INSERT`.

Reply:
no we cannot, the main problem is that the current regionalToComposite does not work with MultiVectors, we would need an overload that does the exact same thing as the current Vector based version...